cmd/registration-tracker/main.go

```diff
@@ -0,0 +1,377 @@
+}
+
+var enabledTransports = map[pb.TransportType]cj.Transport{
+	pb.TransportType_Min:    min.Transport{},
+	pb.TransportType_Obfs4:  obfs4.Transport{},
+	pb.TransportType_Prefix: prefix.DefaultSet(),
+	pb.TransportType_DTLS:   dtls.Transport{},
+}
+
+func (c *transportStats) GetCounts() string {
+
+	return fmt.Sprintf("%d %d %d %d %d %d %d %d %d %d %d %d %d %d %d %d",
+		atomic.LoadInt64(&c.minCount),
+		atomic.LoadInt64(&c.dtlsCount),
+		atomic.LoadInt64(&c.obfs4Count),
+		atomic.LoadInt64(&c.unknownTransportCount),
+		atomic.LoadInt64(&c.prefixCount),
+		atomic.LoadInt64(&c.prefixMinCount),
+		atomic.LoadInt64(&c.prefixGetLongCount),
+		atomic.LoadInt64(&c.prefixPostLongCount),
+		atomic.LoadInt64(&c.prefixHTTPRespCount),
+		atomic.LoadInt64(&c.prefixTLSClientHelloCount),
+		atomic.LoadInt64(&c.prefixTLSServerHelloCount),
+		atomic.LoadInt64(&c.prefixTLSAlertWarningCount),
+		atomic.LoadInt64(&c.prefixTLSAlertFatalCount),
+		atomic.LoadInt64(&c.prefixDNSOverTCPCount),
+		atomic.LoadInt64(&c.prefixOpenSSH2Count),
+		atomic.LoadInt64(&c.prefixUnknownCount),
+	)
+}
+
+func (c *transportStats) IncrementCount(reg *cj.DecoyRegistration) {
+	transport := *reg.TransportPtr
+	switch transport.Name() {
+	case "MinTransport":
+		atomic.AddInt64(&c.minCount, 1)
+	case "dtls":
+		atomic.AddInt64(&c.dtlsCount, 1)
+	case "obfs4":
+		atomic.AddInt64(&c.obfs4Count, 1)
+	case "PrefixTransport":
+		atomic.AddInt64(&c.prefixCount, 1)
+		switch prefix.PrefixID(reg.TransportParams().(*pb.PrefixTransportParams).GetPrefixId()).Name() {
+		// Becuase reg.TransportParams() returns any, we need to cast it to *pb.PrefixTransportParams and
+		// GetPrefixId() to be able to get the prefix name
+		case "Min":
+			atomic.AddInt64(&c.prefixMinCount, 1)
+		case "GetLong":
+			atomic.AddInt64(&c.prefixGetLongCount, 1)
+		case "PostLong":
+			atomic.AddInt64(&c.prefixPostLongCount, 1)
+		case "HTTPResp":
+			atomic.AddInt64(&c.prefixHTTPRespCount, 1)
+		case "TLSClientHello":
+			atomic.AddInt64(&c.prefixTLSClientHelloCount, 1)
+		case "TLSServerHello":
+			atomic.AddInt64(&c.prefixTLSServerHelloCount, 1)
+		case "TLSAlertWarning":
+			atomic.AddInt64(&c.prefixTLSAlertWarningCount, 1)
+		case "TLSAlertFatal":
+			atomic.AddInt64(&c.prefixTLSAlertFatalCount, 1)
+		case "DNSOverTCP":
+			atomic.AddInt64(&c.prefixDNSOverTCPCount, 1)
+		case "OpenSSH2":
+			atomic.AddInt64(&c.prefixOpenSSH2Count, 1)
+		default:
+			atomic.AddInt64(&c.prefixUnknownCount, 1)
+		}
+	default:
+		atomic.AddInt64(&c.unknownTransportCount, 1)
+	}
+}
+
+func (rs *regStats) PrintAndReset() {
+	rs.m.Lock()
+	defer rs.m.Unlock()
+
+	for i, val := range [2]map[uint]*transportStats{rs.v4Registrations, rs.v6Registrations} {
+		ip_ver := 4
+		if i == 1 {
+			ip_ver = 6
+		}
+		for asn, counts := range val {
+			rs.logger.Infof("regtrack (IPv%d): %d %s %s",
+				ip_ver,
+				asn,
+				counts.cc,
+				counts.GetCounts(),
+			)
+		}
+	}
+	rs.v4Registrations = make(map[uint]*transportStats)
```

Comment:
Does this leak the old rs.v4Registrations? I would guess no, but might be good to double check that we don't need to delete the old one